Add atom feed export

TristanCacqueray opened the pull request:
This change implements a feed view for atom reader.

- [x] Add feed content from note doc
- [x] Document the feature

TristanCacqueray committed:
Add feed view

This change enables producing ATOM feed for a given note by adding
the following configuration:

  feed:
    enable: true

TristanCacqueray commented:
Hello @srid, hope you don't mind me trying to implement #394 . I have a couple of questions:

- how to pass the configuration such as feedName, feedUrl, feedQuery, ... ?
- how to get the absolute url of a note?

srid commented:
Hey!

> how to pass the configuration such as feedName, feedUrl, feedQuery, ... ?

Normally, you can put them in either YAML configuration (eg: `index.yaml`) or Markdown frontmatter, depending on the semantics. Are you implementing one global feed? Or support for multiple feeds? 

> how to get the absolute url of a note?

https://github.com/srid/emanote/blob/d1dd7ea4ad89dc74f8d4b90650e07e788b489bb6/emanote/src/Emanote/Route/SiteRoute/Class.hs#L153

EDIT: That gives you relative to root URL, you can pass domain itself in config.

TristanCacqueray commented:
I'd be happy to implement what you see fit. I think a global feed with a single query to get the timeline would be good for most cases.

Thanks, I'll look into adding a global feedQuery to the YAML config.

srid commented:
Multi-feed may be simpler to implement and is also future-proof (what if someone wants to start a second blog?).

Consider an example `blog.md` that has a query to show recent blog listing. Can we automatically produce a corresponding `blog.xml` Atom feed based off that query, if the user has asked for it in the frontmatter? viz:

~~~
$ cat blog.md
---
slug: blog
generate_feed: true
---

## Recent posts

```query
tag:blog
```
~~~

This would be similar https://neuron.zettel.page/feed

TristanCacqueray commented:
Alright, that sounds good. It's just that I don't know yet where to hook the `blog.xml` file when there is a `generate_feed: true` in the note :-)

srid commented:
You can perhaps start with adding a new field of type `Maybe Feed`:

https://github.com/srid/emanote/blob/d1dd7ea4ad89dc74f8d4b90650e07e788b489bb6/emanote/src/Emanote/Model/Note.hs#L39-L45

Then perhaps add a `ResourceRoute_Feed LMLRoute` constructor here:

https://github.com/srid/emanote/blob/d1dd7ea4ad89dc74f8d4b90650e07e788b489bb6/emanote/src/Emanote/Route/SiteRoute/Type.hs#L36-L38

TristanCacqueray commented:
Ok, thanks, I think I can do it that way.

Then I have another question about the html template, how to add a
```
<link href="${feed:feedurl}" type="application/atom+xml" rel="alternate" title="${feed:title}"/>
```
conditionally? It seems like I should add the values in `Emanote.View.Common.commonSplices`, but I'm not sure how to do the `if feed is defined; then add the <link>` in the `base.tpl`. Is this the intended way, or should it be a different template? I could use some help here too.

srid commented:
One approach is to hook into `page.headHtml`, but there may be better approaches (checkout how sidebar is rendered).

https://github.com/srid/emanote/blob/d1dd7ea4ad89dc74f8d4b90650e07e788b489bb6/emanote/default/templates/base.tpl#L22

I suppose we can tackle this latter seeing how the implementation shapes up.

TristanCacqueray commented:
I may have spoken too soon :-) I'm shooting in the dark here as I'm not familiar with GADTs or type families... If you could check the shape of 912127a, it would be good to know if I'm on the right path. It seems like I needed to add a new `'Xml` FileType to please ghc.

So If that looks good to you, then it seems like I can almost pass the note metadata to the renderFeed!

srid commented:
I'll take a look. Meanwhile, you'll find [this post](https://lexi-lambda.github.io/blog/2021/03/25/an-introduction-to-typeclass-metaprogramming/) quite useful to understand the foundations behind those concepts (GADTs, type families).

Also, once you get something working, we can then refactor the code latter.

TristanCacqueray commented:
@srid Thanks!

39ba3dc appears to work with the run command, e.g. adding to my blog.md:

```
---
slug: blog
generate_feed: true
feed_url: https://tristancacqueray.github.io
feed_name: Tristan's Zettelkasten
---

# Blog

My blog articles:

```query {.timeline}
tag:blog
```

I get the feed when running `curl http://127.0.0.1:8080/blog.xml`.
However the xml file is not available when running the `gen` command.

srid commented:
> However the xml file is not available when running the `gen` command.

You want to add your feed URLs to the return value of this function: https://github.com/srid/emanote/blob/d1dd7ea4ad89dc74f8d4b90650e07e788b489bb6/emanote/src/Emanote/Route/SiteRoute/Class.hs#L38-L39

(It is basically Ema's `routeUniverse`: https://ema.srid.ca/guide/route/class)

TristanCacqueray commented:
@srid Thanks! I updated my site with this change (see: https://github.com/TristanCacqueray/TristanCacqueray.github.io/commit/d4b0067968e0a3d97a03da2de35341c81a59ae29), and here is the new feed https://tristancacqueray.github.io/blog.xml :)

srid commented on emanote/src/Emanote/View/Feed.hs:
```diff
@@ -0,0 +1,51 @@
+import Text.Atom.Feed.Export qualified as Export (textFeed)
+
+getNotes :: Text -> Model -> [Note]
+getNotes tagName model = case parseQuery ("tag:" <> tagName) of
```
Instead of hardcoding the query, we should make that configurable in the frontmatter or -- better -- automatically use the first (and only) query in the body of the Markdown note.

srid commented on emanote/src/Emanote/Model/Type.hs:
```diff
@@ -243,6 +245,23 @@ modelLookupNoteByHtmlRoute r =
+isNoteFeed :: Note -> Bool
+isNoteFeed note = case _noteMeta note ^? key "generate_feed" % _Bool of
+  Just True -> True
+  _ -> False
```
Can you make this a bit static, by storing the flag in the `Note` record itself?

srid commented on emanote/src/Emanote/Model/Note.hs:
```diff
@@ -147,6 +150,16 @@ queryNoteTitle r doc meta =
       x
 
+-- | The xml route intended by user for this note.
+noteXmlRoute :: Note -> R 'R.Xml
```
Shouldn't this return a `Maybe`, because not all notes will have a feed?

srid commented on emanote/src/Emanote/Model/Note.hs:
```diff
@@ -147,6 +150,16 @@ queryNoteTitle r doc meta =
+    Nothing ->
+      R.withLmlRoute coerce _noteRoute
+    Just slugs ->
+      R.mkRouteFromSlugs slugs
```
I don't think you need this logic; just call `noteHtmlRoute` and coerce its result.

srid commented on emanote/src/Emanote/Route/Ext.hs:
```diff
@@ -15,6 +15,7 @@ data SourceExt = SourceExt
 data FileType a where
   LMLType :: LML -> FileType SourceExt
   Yaml :: FileType SourceExt
+  Xml :: FileType SourceExt
```
This should be `Xml :: FileType ()` since `.xml` is not a source file (it is rather an output file).

srid commented on emanote/src/Emanote/Route/R.hs:
```diff
@@ -88,6 +88,14 @@ encodeRoute (R slugs) =
   let parts = Slug.unSlug <$> slugs
    in withExt @a @ft $ toString $ T.intercalate "/" (toList parts)
 
+decodeXmlRoute :: FilePath -> Maybe (R 'Xml)
```
Maybe generalize `decodeHtmlRoute` and use that?

TristanCacqueray replied:
I'm not exactly sure how to do that.

srid replied:
Okay, let's address this at the end.

srid commented on emanote/src/Emanote/View/Feed.hs:
```diff
@@ -0,0 +1,51 @@
+      feedSlug <- meta ^? key "slug" % _String
+      feedUrl <- meta ^? key "feed_url" % _String
+      feedName <- meta ^? key "feed_name" % _String
+      let noteUrl note =
```
Instead of,

```yaml
generate_feed: true
feed_url: https://tristancacqueray.github.io
feed_name: Tristan's Zettelkasten
```

let's go with:

```yaml
feed:
  enable: true
  url: https://tristancacqueray.github.io
  title: Tristan's Zettelkasten
```

Actually, what is `url` doing here? Shouldn't it be in `index.yaml`? If you have multiple feeds, they all have a common base URL.

TristanCacqueray replied:
Perhaps the title could default to the _noteTitle too?

TristanCacqueray replied:
It doesn't seem like `index.yaml` already has a `url`, so I think we should fallback to the index only if the feed doesn't have one already. Could you please show how to get the `_noteMeta` of the `index.yaml` ?

srid replied:
index.yaml can contain the defaults, which then gets merged with `.md`'s frontmatter YAML

This function automatically does the merge (using `SData.mergeAesons`):

https://github.com/srid/emanote/blob/84aaaac7384571576506377ccdd5182fc5662ac7/emanote/src/Emanote/Model/Meta.hs#L18-L21

https://github.com/search?q=repo%3Asrid%2Femanote%20lookupRouteMeta&type=code

TristanCacqueray replied:
Alright thanks! So last commit pulls the url from either:

```
feed:
  url: X
```

Or from the index.yaml:

```
site:
  url: X
```

srid reviewed (commented):
Not a *thorough* review, but I thought I'd give a quick feedback.

TristanCacqueray commented:
Thanks for the review! I'll address the remaining comments later.

TristanCacqueray commented:
Hey @srid, thank you for the help, I think this PR is now ready for review. It doesn't seems like there is a test infrastructure I could use to validate the change, but beside that, the feature should be complete. Let me know if there is anything else I could do, or feel free to adapt the implementation to your taste.

TristanCacqueray commented:
> You'll find [this post](https://lexi-lambda.github.io/blog/2021/03/25/an-introduction-to-typeclass-metaprogramming/) quite useful to understand the foundations behind those concepts (GADTs, type families).

Thank you for the link. I think I read it before, but I haven't found a use for this yet. Though it's good to see this in action in emanote :-)

srid commented:
I made a typo in the YAML frontmatter ("enable" instead of "enabled") which caused Emanote to error out. A bug somewhere? I haven't looked in detail:

```diff
diff --git a/docs/guide/query.md b/docs/guide/query.md
index 2af625a8..188d946c 100644
--- a/docs/guide/query.md
+++ b/docs/guide/query.md
@@ -1,5 +1,7 @@
 ---
 order: -95
+feed:
+  enable: true
 ---
 
 # Obsidian-style queries
```

➡️

```
emanote: attempt to encode missing note: LMLRoute_Md R[/guide/query.md]
CallStack (from HasCallStack):
  error, called at src/Relude/Debug.hs:289:11 in relude-1.1.0.0-KCT8nCRmwYB8RNjx4L9ojh:Relude.Debug
  error, called at src/Emanote/Route/SiteRoute/Class.hs:115:10 in main:Emanote.Route.SiteRoute.Class
  encodeResourceRoute, called at src/Emanote/Route/SiteRoute/Class.hs:90:9 in main:Emanote.Route.SiteRoute.Class
  emanoteRouteEncoder, called at src/Emanote.hs:48:27 in main:Emanote
```

srid commented:
http://127.0.0.1:9010/guide/query.xml => `<error>index.yaml or note doesn't have url</error>`

This should cause static-site generation to fail, but currently it does not (because `nix build .#docs` succeeds). The static-site sanity checks are done around here: https://github.com/srid/emanote/blob/34412693d180b9da51109e204d19800f2655843c/emanote/src/Emanote.hs#L88C1-L89

TristanCacqueray commented:
Oops, the error was happening because the discovery link assumed the feed was enabled (it was also triggered by setting `enabled: false`). This should be fixed by e610b8a6.

Then 79bc64ef should make the generation fail on feed error.

srid commented:
There is still an issue here.

@TristanCacqueray Are you using Nix to develop this PR? [The CI failure](https://jenkins.srid.ca/job/Emanote/job/PR-451/18/console) is coming from `nix build .#docs` (I can also see the same error in local dev live server, run via `bin/run`).

<img width="874" alt="image" src="https://github.com/srid/emanote/assets/3998/64a85b35-1355-4a3f-a855-0dab91eef4a8">

TristanCacqueray commented:
@srid arg, I messed up the hlint suggestion, 306ffca should pass the ci.

srid commented:
Great. I'd suggest the following API:

```yaml
feed:
  enable: true  # Note: "enable", not "enabled"
  siteUrl: https://emanote.srid.ca 
````

srid commented:
To clarify, you don't need `site.url`. If you use `getEffectiveRouteMeta`, looking up `feed.siteUrl` will automatically search up to the top-level `index.yaml` to find it.

TristanCacqueray commented:
@srid alright, good suggestions. Last commit should implement the change.

srid commented:
I found a couple of issues

1. Adding `feed.title` seems to have no effect, as the feed is still using the note title. 1d1d31f
2. It seems you can't link to the .xml in Markdown, emanote thinks it is broken link. Looks like the route decoder is not properly implemented? 2157e26

TristanCacqueray commented:
1. should be fixed in b4c7ffa
2. It seems like the culprit comes from `mkModelRouteFromFilePath`. Do I need to add a ModelRoute_Feed constructor to ModelRoute ?

srid commented:
> 2\. It seems like the culprit comes from `mkModelRouteFromFilePath`. Do I need to add a ModelRoute_Feed constructor to ModelRoute ?

Yea.

But this is making me think that if we probably should refactor the output route type as follows:

```haskell
data ResourceRoute
  = ResourceRoute_StaticFile StaticFileRoute FilePath
  | ResourceRoute_LML LMLRoute LMLView

data LMLView = LMLView_Html | LMLView_Atom
```

Then you can just reuse `LMLView` in the `ModelRoute`.

TristanCacqueray commented:
What about LMLRoute, do you think we need a `LMLRoute_Xml (R 'Xml)` constructor?

TristanCacqueray commented:
Note sure how to get the link to point at the actual xml resource, but e7484b9 fixes the link render error, but it now points at the html file. It seems like we need a new LMLRoute, but could you confirm?

TristanCacqueray commented:
@srid thinking about it, changing the ModelRoute to accommodate feed links doesn't fit very well. In the last commit 2815a24 I propose a different strategy, would that work for you?

TristanCacqueray commented:
@srid I've squashed the commits and removed the ModelRoute change (e.g. internal link to feed xml are not working). Would it be possible to do that in a followup, since the feature is working otherwise.

srid commented:
@TristanCacqueray It is best to fix this glitch before merging to `master`. We don't want surprises; in fact when I was testing your branch locally right now, I was surprised that the `query.xml` took me to some other domain.

 I've been meaning to do this myself, so I just took a brief look at it. I think you can solve this problem by replacing `(Either MN.Note SF.StaticFile)` in Resolve.hs with `(Either (MN.Note, LMLView) SF.StaticFile)` (or equivalent). And maybe `ModelRoute_LML LMLView LMLRoute` constructor as well in `ModelRoute`. Haven't had the time to take a deeper look ...

srid committed:
Merge branch 'master' into feed

srid committed:
make static links work

also contains vscode auto formatting (meh)

srid committed:
changelog

srid committed:
fmt

srid commented:
> And maybe `ModelRoute_LML LMLView LMLRoute` constructor as well in `ModelRoute`.

Just implemented this.

Aside: I also realized that it has become a pain to work with Emanote code. I find that it has gotten way too complex! Time for a rewrite? 🤔

srid committed:
up ver

TristanCacqueray commented:
@srid thank you for your help and for merging this feature! For what it's worth, I don't think Emanote is too complex, perhaps it deserves a rewrite, but can it be less complicated? My next wish is a better timeline as requested in [#422](https://github.com/srid/emanote/discussions/422), for example to show a reading time estimation based on the word count.
